refactor(docs-infra): use eslint in aio's example-lint script

dario-piotrowicz opened the pull request:
Instead of the deprecated tslint use eslint in the aio's example-lint script

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
Linting for the aio example is done using tslint (which is deprecated)

Issue Number: N/A


## What is the new behavior?
Linting for the aio example is done using eslint

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

dario-piotrowicz commented:
@gkalpak here we are! :smile: :tada: 

But actually.... eslint seems to not handling very well the large amount of files to check here, and it is so much slower than tslint in this instance... whilst this check takes tslint around 40 seconds on my machine, eslint takes around 4 minutes.... :disappointed: 
(and you can see I am ignoring all possible files except the ts ones)

This is rather unfortunate... but I guess it would still make sense to move away from the deprecates tslint.... what do you think?

dario-piotrowicz commented on aio/.eslintrc.json:
```diff
@@ -58,7 +58,7 @@
         "no-tabs": "error",
         "no-underscore-dangle": "off",
-        "no-unused-expressions": "error",
+        "no-unused-expressions": "off",
```
I am removing this (in general not just for the examples) because I noticed it should really be of as the typescript one is already applied, as you can for example see here:
![Screenshot at 2021-08-20 20-53-17](https://user-images.githubusercontent.com/61631103/130294804-4381f9f7-4eef-48d4-a89c-1dad4d51a4ac.png)

I think it could be a good idea to also explicitly add the typescript rule with `"error"`, but it's not necessary I guess, what do you think?

dario-piotrowicz commented on aio/content/examples/.eslintrc.json:
```diff
@@ -0,0 +1,13 @@
+    "extends": "../../.eslintrc.json",
+    "ignorePatterns": ["*.*", "node_modules/", "!src/app/**/*.ts", "styleguide/**/*.avoid.ts"],
+    "rules": {
+        "@angular-eslint/component-selector": "off",
```
as the examples are generic it makes sense they don't use the standard `aio` prefix

dario-piotrowicz commented on aio/content/examples/.eslintrc.json:
```diff
@@ -0,0 +1,13 @@
+        "@angular-eslint/directive-selector": "off",
+        "@angular-eslint/no-inputs-metadata-property": "off",
+        "@angular-eslint/no-outputs-metadata-property": "off",
+        "@angular-eslint/template/click-events-have-key-events": "off",
+        "@angular-eslint/template/accessibility-alt-text": "off",
+        "@angular-eslint/template/accessibility-label-has-associated-control": "off"
```
I needed to add this in because I tried adding the disabling comments for these rules in the ts files but that didn't seem to work unfortunately (maybe some bug in eslint?)

rafaelss95 replied:
See if https://github.com/angular/angular/pull/43218#discussion_r693282000 helps with that.

rafaelss95 replied:
Have you tried to disable it in the template metadata instead? For example:

```typescript
@Component({
  template: `
    <!-- eslint-disable-next-line @angular-eslint/template/click-events-have-key-events -->
    <a (click)="onClick()"></a>
  `
})
export class Test {}
```

dario-piotrowicz replied:
Yes you are correct that putting the comment in the template string actually works fine but that is not an option unfortunately as it would show in the example on the website

For example:
![Screenshot at 2021-08-21 11-01-57](https://user-images.githubusercontent.com/61631103/130318401-ebcec13d-926a-4079-9011-765b9a6c55aa.png)

dario-piotrowicz replied:
I'm not sure :thinking:

may be difficult as the template is specified as a string literal, and may actually be interpreted as a string itself (for example indeed I can't see any syntax highlighting in the screenshot above)

I am also not sure if that would be a good addition as maybe we could want to show such comments in some instances

Either way this seems to me like quite out of scope for this PR, maybe I can open an issue to discuss it further :slightly_smiling_face: 

@gkalpak any thoughts on this?

gkalpak replied:
I think it makes sense to have the tooling remove linting comments.

That said, if there are many examples that need disabling the `accessibility-label-has-associated-control` rule, then it makes sense to globally disable it. Otherwise, let's keep the rule turned on and disable is specific places.

dario-piotrowicz replied:
ok I see :slightly_smiling_face: , but making the change in the tooling seems quite out of scope here... also to be totally honest I would not even know where to start, it doesn't really look like a quick/simple change... 

@gkalpak do you agree? what course of action would you suggest here?

gkalpak replied:
If this PR is adding new comments in HTML code snippets (that will be confusing for readers), we should update the tooling to be able to remove such comments before merging this PR. (Does that make sense?)

As far as implementing this, you don't have to worry about it. Someone from the `docs-infra` team will take care of it (because it requires familiarity with the tooling).

dario-piotrowicz replied:
@gkalpak yes that makes total sense! but, I am not 100% we are on the same page (sorry if the screenshot above may have misled you, I just sent it to show the effect those comments would have)

In the PR there are currently no such comments, those rules are being set off in the eslintrc and not disabled in the html snippets shown in the aio application! (lol it would indeed look pretty crappy to have eslint-disable comments sprinkled amongst the html code samples :sweat_smile:)

So this to me doesn't seem like a change in the tooling absolutely needed here and now, but something that I would consider a nice-to-have/future improvement.

(unless you consider the @angular-eslint/template rules necessary to have)

(ps: I think it's also worth mentioning that this are additional rules compared to the current tslint implementation, so turning them off doesn't make things worse, it just sort of doesn't make them better)

dario-piotrowicz commented on aio/content/examples/component-interaction/src/app/votetaker.component.ts:
```diff
@@ -18,7 +18,11 @@ export class VoteTakerComponent {
   voters = ['Narco', 'Celeritas', 'Bombasto'];
 
   onVoted(agreed: boolean) {
-    agreed ? this.agreed++ : this.disagreed++;
```
this caused an unused expression lint error, I could disable the lint rule but I actually think it makes sense to just use an if statement

dario-piotrowicz commented on aio/content/examples/dependency-injection-in-action/src/app/logger.service.ts:
```diff
@@ -13,6 +13,10 @@ export class LoggerService {
 
   private log(msg: any, isErr = false) {
     this.logs.push(msg);
-    isErr ? console.error(msg) : console.log(msg);
```
as above this caused an unused expression error so I used changed it into an if statement instead of disabling the rule

dario-piotrowicz commented on aio/content/examples/dependency-injection/e2e/src/app.e2e-spec.ts:
```diff
@@ -132,7 +132,7 @@ describe('Dependency Injection Tests', () => {
 
     it('should have button', async () => {
       expect(await element.all(by.cssContainingText('button', 'Next User'))
-        .get(0).isDisplayed()).toBe(true, '\'Next User\' button should be displayed');
```
some like this didn't actually cause lint issues but I thought it was cleaner and more consistent with other parts to avoid the escaped single quotes, I hope this is ok

dario-piotrowicz replied:
I wanted to add that not escaping single quotes is necessary in templates because I have seen that if a template has an escaped quote then eslint errors on that file (with something like: `preprocess: ERROR could not parse @Component() metadata`)

I believe it's some sort of bug in eslint / angular-eslint

I don't think this is an issue but I just felt compelled to point that out so that you know :slightly_smiling_face: 

___

> 
>  also note that there isn't even an occurrence of that in the current examples, I only noticed it because the automated eslint fix converted a couple of templates defined in backticks to single ticks
> for example, it converted something like:
>  ```ts
>  template: `<p>don't</p>`
>  ```
>  into: 
>  ```ts
>  template: '<p>don\'t</p>'
>  ```
> which generated an error during linting and which I fixed by converting to:
> ```ts
> template: "<p>don't</p>"
> ```

rafaelss95 replied:
I guess this is related to https://github.com/angular-eslint/angular-eslint/issues/207

rafaelss95 replied:
I guess this is related to https://github.com/angular-eslint/angular-eslint/issues/207.

dario-piotrowicz replied:
Yes exactly :+1:

dario-piotrowicz replied:
@JamesHenry Thank you so very much for your reply!!! (https://github.com/angular/angular/pull/43218#issuecomment-923010590) :slightly_smiling_face: 

It's so appreciated that you remembered this PR and wen't back to let us know!!!! thanks!!! :heart_eyes: 

I am commenting here as I believe that it is indeed here that we discussed the thing initially :slightly_smiling_face: 

I actually think that the workaround her was very clean and just consisted in not having backslashes in the component's templates, so I think it is ok just to keep it like it is now :slightly_smiling_face:  (@gkalpak do you agree?)

But we can definitely bump the package version so that in the future the problem doesn't come back even if there are backslashes in the templates :slightly_smiling_face::+1:

dario-piotrowicz commented on aio/content/examples/testing/src/app/demo/demo.ts:
```diff
@@ -1,4 +1,4 @@
-// tslint:disable: directive-selector forin no-input-rename
+/* eslint-disable guard-for-in, @angular-eslint/no-input-rename, @angular-eslint/template/click-events-have-key-events */
```
disabling template rules doesn't work, sorry I need to remove that one here

rafaelss95 replied:
Hey @dario-piotrowicz, have you tried to disable it under the template instead? For example:

```typescript
@Component({
  template: `
    <!--  eslint-disable-next-line @angular-eslint/template/click-events-have-key-events -->
    <a (click)="onClick()"></a>
  `
})
export class Test {}
```

dario-piotrowicz commented on aio/content/examples/universal/src/app/app.module.ts:
```diff
@@ -45,7 +45,7 @@ import { isPlatformBrowser } from '@angular/common';
 export class AppModule {
   constructor(
-    @Inject(PLATFORM_ID) private platformId: object,
+    @Inject(PLATFORM_ID) private platformId: Record<string, unknown>,
```
`object` is banned, so I put `Record` instead (as it was suggested by eslint), there's probably a better alternative to this, I will have a think of it

dario-piotrowicz replied:
no, `isPlatformBrowser` below required an `object` as it's argument so I think this either needs to be a `Record` as I did or `any` :disappointed: 

(or we can just disable the rule here)

(
ps: `unknown` doesn't work
![Screenshot at 2021-08-21 12-03-37](https://user-images.githubusercontent.com/61631103/130319780-abc0132c-2300-445d-a1e2-df05418c9d48.png)
)

dario-piotrowicz replied:
`Object` capitalized is even worse I believe:
![Screenshot at 2021-09-02 22-15-01](https://user-images.githubusercontent.com/61631103/131917358-f5c18ac0-ea8e-47d6-a316-8b1258d580cc.png)

With `object` we get:
![Screenshot at 2021-09-02 22-19-53](https://user-images.githubusercontent.com/61631103/131917697-2567d5dc-44ce-4969-a01e-144f34eeb7d4.png)

so I guess I can just use `object` and disable the rule?

> Wow the Parameters utility-type is a nice trick I didn't know of it to be honest! :slightly_smiling_face:  still yeah it doesn't look much more readable :disappointed:

dario-piotrowicz replied:
I also wanted to mention that this example gets downloaded from the docs:
![Screenshot at 2021-09-02 22-27-14](https://user-images.githubusercontent.com/61631103/131918677-803b629c-dec8-4933-ae1d-6924061d6396.png)

so maybe I would imagine it would be better to disable the rule for the whole file instead of dropping an `eslint-disable-next-line` comment there in the middle of the code

dario-piotrowicz replied:
So @gkalpak do I use `object` and simple put a `/* eslint-disable @typescript-eslint/ban-types */` at the top of the file? would you be ok with that? :slightly_smiling_face:

gkalpak replied:
I am torn between `object` and `Object`. `object` is generally preferable, but the `PLATFORM_ID` token is typed as `Object` (🤷).
Leaving it up to you which type to use from the two. Regarding disabling the rule, I would prefer to use the `eslint-disable-next-line` variant.

dario-piotrowicz replied:
I don't have a strong opinion on the type matter (the only thing that I have an opinion on was having the rule at the top, which anyway I won't go for :sweat_smile:) given the fact that we are not sure which to choose I would simple go with the solution with less diff and keep it as `object` :slightly_smiling_face:

dario-piotrowicz replied:
Change made ([commit](https://github.com/angular/angular/pull/43218/commits/a6d945b45ea7de2db2e3a601afb6922fe45f1f47)) I hope it's fine :slightly_smiling_face:

dario-piotrowicz commented on aio/content/examples/.eslintrc.json:
```diff
@@ -0,0 +1,13 @@
+    "ignorePatterns": ["*.*", "node_modules/", "!src/app/**/*.ts", "styleguide/**/*.avoid.ts"],
+    "rules": {
+        "@angular-eslint/component-selector": "off",
+        "@angular-eslint/directive-selector": "off",
```
also this one seemed often not respected (since this are examples I'd imagine)

dario-piotrowicz replied:
sorry I totally dropped the ball on this one :bow: :bow: :bow: :bow: :bow: :bow: :bow: :bow: :bow: 

___

Added in this [commit](https://github.com/angular/angular/pull/43218/commits/8ee943a25a60ae4515e6b616e849b6cd0752b182), please have a look I had to disable the rule in some files because they weren't using the prefix (not sure why it wasn't being picked up by `tslint` :thinking:)

rafaelss95 commented on aio/content/examples/dependency-injection/src/app/test.component.ts:
```diff
@@ -22,8 +22,8 @@ export class TestComponent {
-  const expectedHeroes = [{name: 'A'}, {name: 'B'}]
-  const mockService = <HeroService> {getHeroes: () => expectedHeroes }
+  const expectedHeroes = [{name: 'A'}, {name: 'B'}];
+  const mockService = <HeroService> {getHeroes: () => expectedHeroes };
```
I think that if you do the following you don't have to disable the `@typescript-eslint/consistent-type-assertions` rule:

```suggestion
  const mockService = {getHeroes: () => expectedHeroes } as HeroService;
```

dario-piotrowicz replied:
yeah, you're totally right, sorry about this :sweat: 

(and thanks for noticing it! :slightly_smiling_face:)

rafaelss95 commented on aio/content/examples/form-validation/src/app/reactive/hero-form-reactive.component.ts:
```diff
@@ -1,4 +1,4 @@
-/* tslint:disable: member-ordering forin */
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
```
There are so many instances of disabling this rule that I think it might be better not to enable this rule at all.

dario-piotrowicz replied:
yeah, that's a good point, but the other rules I disabled (except the template ones) were rules that made sense not to follow for the examples (for example the use of the `aio` prefix), but this one I am not sure if has a reason not to be followed, maybe if could be left enabled to discourage people to use the assertion? :thinking:

gkalpak replied:
I also think it makes sense to globally disable the rule, since there are so many violations already.
(If `.eslintrc.json` supports comments, you can add a TODO comment to fix violations and enable the rule eventually.)

dario-piotrowicz replied:
ok will do! :+1:

rafaelss95 commented on aio/content/examples/pipes/src/app/flying-heroes.pipe.ts:
```diff
@@ -1,4 +1,4 @@
-/* tslint:disable use-pipe-transform-interface */
+/* eslint-disable @angular-eslint/use-pipe-transform-interface */
```
You could just implement `PipeTransform` and remove this disable comment.

dario-piotrowicz replied:
This to me just seems like a bug in eslint :cry: 

This is the code:
![Screenshot at 2021-08-21 11-33-16](https://user-images.githubusercontent.com/61631103/130319074-4f90261a-f004-4657-8389-5b9be4ef02a9.png)

As you can see the example is practically already implementing the `PipeTransform` by extending the`FlyingHeroesPipe` class

Also note that this code is shown on the website so I really would not want to add noise to it by implementing directly the `PipeTransform`:
![Screenshot at 2021-08-21 11-36-05](https://user-images.githubusercontent.com/61631103/130319156-bebcad52-1d0f-403f-8e01-05f90f230aea.png)

rafaelss95 replied:
Ah, I see your point. And just to clarify: it isn't a bug in `@angular-eslint/use-pipe-transform-interface`, it's just because that rule doesn't requires type checking, so we don't have a way to know that it's already implementing `PipeTransform`.

dario-piotrowicz replied:
mh.... sorry I am not sure I follow...why isn't that a bug? shouldn't then the rule require type checking then? :thinking:

rafaelss95 commented on aio/content/examples/testing/src/app/dashboard/dashboard-hero.component.ts:
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable @angular-eslint/template/click-events-have-key-events */
```
I imagine that this and other disable comments like this has no effect at this specific location, right?

dario-piotrowicz replied:
Yes indeed, thanks for noticing this one, I forgot to remove it :sweat:

rafaelss95 commented on aio/content/examples/upgrade-module/src/app/a-to-ajs-providers/app.module.ts:
```diff
@@ -1,4 +1,5 @@
-declare var angular: angular.IAngularStatic;
+/* eslint-disable @angular-eslint/use-lifecycle-interface */
```
Wouldn't it be better to just implement `DoBootstrap` instead (here and in the other `.module.ts` files)?

dario-piotrowicz replied:
yeah, that's a good point, I can do that without problems

But these are shown on the website:
![Screenshot at 2021-08-21 11-57-23](https://user-images.githubusercontent.com/61631103/130319619-8883693e-8333-44ba-8c72-3b9d8f0a5624.png)

So I am just wondering if they are not implementing the interface for a specific reason....
(for example to keep things minimal/simple)

@gkalpak any thoughts? :slightly_smiling_face:

dario-piotrowicz replied:
done :slightly_smiling_face:  ( [commit](https://github.com/angular/angular/pull/43218/commits/45783702ea447e639b9ffcb17c6e22d74a6738b8) )

rafaelss95 commented on aio/content/examples/component-interaction/src/app/votetaker.component.ts:
```diff
@@ -18,7 +18,11 @@ export class VoteTakerComponent {
 
   onVoted(agreed: boolean) {
-    agreed ? this.agreed++ : this.disagreed++;
+    if(agreed) {
```
I think you're missing a whitespace in this and in `logger.service.ts`: `if(isErr)` => `if (isErr)`.

dario-piotrowicz commented:
@rafaelss95 Thank you so very much for your nice reviewing! it was really helpful and much appreciated! :smiley: 

I replied to your comments feel free to have another look :slightly_smiling_face: 

(ps: again same as list time, for some reason I cannot re-request your review :confused:)

rafaelss95 commented on aio/content/examples/upgrade-module/src/app/downgrade-io/hero-detail.component.ts:
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable @angular-eslint/template/accessibility-label-has-associated-control */
```
You can remove it from here and from `downgrade-static/hero-detail.component.ts` 😄

dario-piotrowicz replied:
yeah sorry! :sweat::+1:

rafaelss95 commented on aio/content/examples/dependency-injection-in-action/src/app/runners-up.ts:
```diff
@@ -9,16 +9,16 @@ export const RUNNERS_UP = new InjectionToken<string>('RunnersUp');
+          .join(', ')
 // #docregion factory-synopsis
-  };
+  ;
```
Is this semicolon intended to be here alone? 😆

dario-piotrowicz replied:
good catch :sweat_smile::+1:

dario-piotrowicz commented:
Hi @JamesHenry , thank you so very much for your awesome comment and also for the great contribution you've done to the community with your `eslint` packages ( needless to say they are awesome :slightly_smiling_face: )

I must admit I didn't really read that part of the readme :sweat: 
Anyways thanks a lot for the awesome suggestions, this looks really really great!!! :slightly_smiling_face: 

Seems like going with the last alternative would be the best thing for me as it really looks so much faster, but I am not totally sure. The fact is that if in the future we'd need to add new rules with type-checking we would have to change the eslintrc file and potentially fall back into the original time issue

Whilst with the second solution the linting would not be as fast but we would be more future-proof...

Did I get it right @JamesHenry ? that's you're view on this what would you go for?

Also since I am just a simple contributor I would really love to hear what someone from the angular team would have to say about this :slightly_smiling_face: (@gkalpak I'm counting on you! :smile:)

gkalpak reviewed (changes requested):
As usual, great work there, @dario-piotrowicz :100:
And thx for chiming in on performance, @JamesHenry :heart:

I've left a few comments. I think I have addressed all your questions (let me know in case I've missed something). Regarding speed, I suggest we first focus on getting the linting behavior we want and then look into making it fast.

Based on prior discussions (here and during the docs-infra team meeting), here is some additional work than needs to be done (not necessarily as part of this PR) before landing this PR:

- [x] Remove `aio/content/examples/tslint.json` (and the `tslint` dependency if not needed any more). Determine if we still need `aio/content/examples/tsconfig.json`?
- [x] Ensure we strip off files/config related to linting (such as `tslint.json` files) from the generated StackBlitz examples and ZIP archives.
- [x] Have tooling remove linting comments from code snippets.

(You don't have to worry about the tasks listed above, @dario-piotrowicz. I am just putting them here for tracking purposes.)

dario-piotrowicz commented on aio/content/examples/component-overview/src/app/app.component.spec.ts:
```diff
@@ -16,7 +16,7 @@ describe('AppComponent', () => {
   });
 
-  it(`should have as title 'component-overview'`, () => {
+  it("should have as title 'component-overview", () => {
```
wow good spot! :eyes:

dario-piotrowicz commented on aio/content/examples/dependency-injection/src/app/heroes/hero.service.provider.ts:
```diff
@@ -5,13 +5,11 @@ import { Logger } from '../logger.service';
-const heroServiceFactory = (logger: Logger, userService: UserService) => {
-  return new HeroService(logger, userService.user.isAuthorized);
-};
+const heroServiceFactory = (logger: Logger, userService: UserService) => new HeroService(logger, userService.user.isAuthorized);
```
I'll have a look I have seen the linter complaining about line length, I think this may just not yet reach the limit

dario-piotrowicz replied:
Ok I checked that and we currently have a `max-len` of 140 (we are not setting it I think it must be some sort of default value of eslint, I could not fine much info online about it), the line in question is 129 so it is a bit shorter than the max we allow, let me know what you'd like me to do :slightly_smiling_face: 

> also note that we seem to use 140 as a max-len value across the board as you can see by searching the tslint equivalent rule: https://github.com/angular/angular/search?p=2&q=%22max-line-length%22

gkalpak replied:
Ideally, we should have a 100 limit length, but if that causes too many violations it is not fixing in this PR.
Maybe we could aim for 120 for now (if there are not too many violations)? Otherwise, we can keep the current default limit of 140 and do it in a future PR (or not :wink:).

dario-piotrowicz replied:
When you suggest about limits I image we are talking about `aio` in general, not just the examples, am I right?

With anything below 120 the number of errors becomes quite significant, with 120 we get:
 - 70 errors in aio
 - 0 in docs
 - 2 in examples
 
I'll set that then, I think it's manageable :slightly_smiling_face::+1:

dario-piotrowicz replied:
Change made ( [commit](https://github.com/angular/angular/pull/43218/commits/df01d5ab8d97ff0471d5a7a9cec3e6d5a9915339) ) I hope it's fine :slightly_smiling_face:

gkalpak replied:
I was talking about the examples for now, but...there's no holding you back 😃

dario-piotrowicz commented on aio/content/examples/hierarchical-dependency-injection/src/app/villains.service.ts:
```diff
@@ -2,7 +2,7 @@ import { Injectable } from '@angular/core';
 import { of } from 'rxjs';
 
-export interface Villain { id: number; name: string; }
+export interface Villain { id: number, name: string }
```
yes totally with you on this! :slightly_smiling_face:

dario-piotrowicz commented on aio/content/examples/pipes/src/app/heroes.ts:
```diff
@@ -1,4 +1,4 @@
-export interface Hero { name: string; canFly: boolean; }
+export interface Hero { name: string, canFly: boolean }
```
again totally agree :slightly_smiling_face::+1:

dario-piotrowicz commented:
> As usual, great work there, @dario-piotrowicz 
> And thx for chiming in on performance, @JamesHenry 
> 
> I've left a few comments. I think I have addressed all your questions (let me know in case I've missed something). Regarding speed, I suggest we first focus on getting the linting behavior we want and then look into making it fast.
> 
> Based on prior discussions (here and during the docs-infra team meeting), here is some additional work than needs to be done (not necessarily as part of this PR) before landing this PR:
> 
> * [ ]  Remove `aio/content/examples/tslint.json` (and the `tslint` dependency of not needed any more). Determine if we still need `aio/content/examples/tsconfig.json`?
> * [ ]  Ensure we strip off files/config related to linting (such as `tslint.json` files) from the generated StackBlitz examples and ZIP archives.
> * [ ]  Have tooling remove linting comments from code snippets.
> 
> (You don't have to worry about the tasks listed above, @dario-piotrowicz. I am just putting them here for tracking purposes.)

Thanks a lot @gkalpak :slightly_smiling_face: 

About the tasks I would indeed suggest not to have them as part of this PR, but if you want you know I would me more than happy to pick them up and help out if I can! :smiley:

mary-poppins commented:
You can preview 608a306 at https://pr43218-608a306.ngbuilds.io/.

mary-poppins commented:
You can preview 89fe51b at https://pr43218-89fe51b.ngbuilds.io/.

mary-poppins commented:
You can preview 4578370 at https://pr43218-4578370.ngbuilds.io/.

dario-piotrowicz commented:
@gkalpak Thanks a lot for the nice reviewing :smile: , I have replied to your comments and/or did the fixes, please check the PR out again :slightly_smiling_face: 
> ps: I have been using fixup commits here as you suggested to me, hopefully this will make your reviewing easier :slightly_smiling_face: 

when you're happy with we would still need to decide what solution to take about performance :slightly_smiling_face::+1:

gkalpak commented on aio/content/examples/template-syntax/src/app/click.directive.ts:
```diff
@@ -1,4 +1,8 @@
-/* tslint:disable directive-selector directive-class-suffix */
+/* eslint-disable @angular-eslint/no-inputs-metadata-property,
```
Is the `@angular-eslint/no-inputs-metadata-property` necessary here? There don't seem to be any inputs used 😕

dario-piotrowicz replied:
yes, it's not necessary, must have ended up here because of a copy-paste, sorry :bow:

gkalpak commented on aio/content/examples/template-syntax/src/app/click.directive.ts:
```diff
@@ -1,4 +1,8 @@
+                  @angular-eslint/no-outputs-metadata-property,
+                  @angular-eslint/no-output-rename,
+                  @angular-eslint/directive-selector,
+                  @angular-eslint/directive-class-suffix */
```
Nit: Please order rules alphabetically 😇  (here ant elsewhere 🙏)

dario-piotrowicz replied:
sorry I will check that! :sweat:

dario-piotrowicz replied:
done ( [commit](https://github.com/angular/angular/pull/43218/commits/e07e314de512efa2f7c7f5db23a65cf500abc4e9) ), it seems it was not ordered alphabetically only in other 2 places, hopefully I didn't miss any :slightly_smiling_face:

gkalpak commented on aio/content/examples/upgrade-module/src/app/a-to-ajs-providers/app.module.ts:
```diff
@@ -1,5 +1,5 @@
-declare var angular: angular.IAngularStatic;
-import { NgModule } from '@angular/core';
+declare let angular: angular.IAngularStatic;
```
```suggestion
declare const angular: angular.IAngularStatic;
```

gkalpak commented on aio/content/examples/upgrade-module/src/app/a-to-ajs-transclusion/app.module.ts:
```diff
@@ -1,5 +1,5 @@
-declare var angular: angular.IAngularStatic;
-import { NgModule } from '@angular/core';
+declare let angular: angular.IAngularStatic;
```
```suggestion
declare const angular: angular.IAngularStatic;
```

gkalpak commented on aio/content/examples/upgrade-module/src/app/ajs-a-hybrid-bootstrap/app.module.ts:
```diff
@@ -1,6 +1,6 @@
-declare var angular: angular.IAngularStatic;
+declare let angular: angular.IAngularStatic;
```
```suggestion
declare const angular: angular.IAngularStatic;
```

gkalpak commented on aio/content/examples/upgrade-module/src/app/ajs-to-a-projection/app.module.ts:
```diff
@@ -1,5 +1,5 @@
-declare var angular: angular.IAngularStatic;
-import { NgModule } from '@angular/core';
+declare let angular: angular.IAngularStatic;
```
```suggestion
declare const angular: angular.IAngularStatic;
```

gkalpak commented on aio/content/examples/upgrade-module/src/app/ajs-to-a-providers/app.module.ts:
```diff
@@ -1,5 +1,5 @@
-declare var angular: angular.IAngularStatic;
-import { NgModule } from '@angular/core';
+declare let angular: angular.IAngularStatic;
```
```suggestion
declare const angular: angular.IAngularStatic;
```

gkalpak commented on aio/content/examples/upgrade-module/src/app/downgrade-io/app.module.ts:
```diff
@@ -1,5 +1,5 @@
-declare var angular: angular.IAngularStatic;
-import { NgModule } from '@angular/core';
+declare let angular: angular.IAngularStatic;
```
```suggestion
declare const angular: angular.IAngularStatic;
```

gkalpak commented on aio/content/examples/upgrade-module/src/app/downgrade-static/app.module.ts:
```diff
@@ -1,5 +1,5 @@
-declare var angular: angular.IAngularStatic;
-import { NgModule } from '@angular/core';
+declare let angular: angular.IAngularStatic;
```
```suggestion
declare const angular: angular.IAngularStatic;
```

gkalpak commented on aio/content/examples/upgrade-module/src/app/upgrade-io/app.module.ts:
```diff
@@ -1,5 +1,5 @@
-declare var angular: angular.IAngularStatic;
-import { NgModule } from '@angular/core';
+declare let angular: angular.IAngularStatic;
```
```suggestion
declare const angular: angular.IAngularStatic;
```

gkalpak commented on aio/content/examples/upgrade-module/src/app/upgrade-static/app.module.ts:
```diff
@@ -1,5 +1,5 @@
-declare var angular: angular.IAngularStatic;
-import { NgModule } from '@angular/core';
+declare let angular: angular.IAngularStatic;
```
```suggestion
declare const angular: angular.IAngularStatic;
```

gkalpak reviewed (changes requested):
Thx for making the changes, @dario-piotrowicz :+1:
(The fixup commits make reviewing much easier btw, thx ❤️)

I've left a few more minor comments. Once these (and other pending comments - including https://github.com/angular/angular/pull/43218#discussion_r701114724) are addressed, we can do some investigation and decide what to do about performance.

I can take a look at the tasks listed in https://github.com/angular/angular/pull/43218#pullrequestreview-744253351 (since they need to be taken care of before merging this PR).

dario-piotrowicz commented:
> Thx for making the changes, @dario-piotrowicz 
> (The fixup commits make reviewing much easier btw, thx )
> 
> I've left a few more minor comments. Once these (and other pending comments - including [#43218 (comment)](https://github.com/angular/angular/pull/43218#discussion_r701114724)) are addressed, we can do some investigation and decide what to do about performance.
> 
> I can take a look as the tasks listed in [#43218 (review)](https://github.com/angular/angular/pull/43218#pullrequestreview-744253351) (since they need to be taken care of before merging this PR).

Thank you very much! :slightly_smiling_face: 

I will check out your new comments right away :slightly_smiling_face: 

About [#43218 (comment)](https://github.com/angular/angular/pull/43218#discussion_r701114724)) as I said I am not sure I can tackle it, and if it can/should be tackled as part of this PR, please let me know what you think (do you think there's an easy way to accomplish that? :thinking:)

About the tasks in [#43218 (review)](https://github.com/angular/angular/pull/43218#pullrequestreview-744253351) I thought I misunderstood, I didn't think they needed to be done before this PR can be merged, and I am not completely sure why this PR couldn't be merged without them, I don't see a cut clear dependency here... but surely if that's what you decided it must be the most appropriate way :slightly_smiling_face::+1:

dario-piotrowicz commented on aio/content/examples/.eslintrc.json:
```diff
@@ -0,0 +1,26 @@
+    "ignorePatterns": ["*.*", "node_modules/", "!src/app/**/*.ts", "styleguide/**/*.avoid.ts"],
+    "rules": {
+        "@angular-eslint/component-selector": [
+            "error",
+            {
+              "type": "element",
+              "prefix": "", // TODO: Fix the code and change the prefix to `"app"` (or whatever makes sense).
+              "style": "kebab-case"
+            }
+          ],
+          "@angular-eslint/directive-selector": [
+            "error",
+            {
+              "type": "attribute",
+              "prefix": ["app", "tho"],
+              "style": "camelCase"
+            }
+          ],
```
@gkalpak sorry :sob: (also the toh :man_facepalming:)

About indentation I am afraid it's because I rely too much on auto formatting tools on a day to day basis :sweat:

dario-piotrowicz commented on aio/content/examples/.eslintrc.json:
```diff
@@ -0,0 +1,26 @@
+        "@angular-eslint/template/accessibility-alt-text": "off",
+        "@angular-eslint/template/accessibility-label-has-associated-control": "off",
+        "@typescript-eslint/no-non-null-assertion": "off" // TODO: Fix the code violating this rule and enable it.
+    }
```
I am so sorry, I remembered this from the other PR, I forgot it here :sweat:

dario-piotrowicz replied:
Done

![Screenshot at 2021-09-03 23-08-42](https://user-images.githubusercontent.com/61631103/132069549-e5a405e7-a955-4629-ad46-6fcb0af7ab66.png)

the order is basically:
 - component-selector
 - directive-selector
 - no-non-null-assertion
 - template/accessibility-alt-text
 - template/accessibility-label-has-associated-control
 - template/click-events-have-key-events
 
 I hope I got it right :sweat_smile:

gkalpak replied:
Perfect ✨

dario-piotrowicz commented on aio/src/app/custom-elements/contributor/contributor.component.ts:
```diff
@@ -11,20 +11,22 @@ import { CONTENT_URL_PREFIX } from 'app/documents/document.service';
-                        <mat-icon class="link-icon">link</mat-icon>
-                    </a>
-                </div>
+            <div class="contributor-image"
```
Note: I checked the collaborators page and this change seems not to have broken anything

dario-piotrowicz commented on aio/src/app/layout/mode-banner/mode-banner.component.ts:
```diff
@@ -4,11 +4,15 @@ import { VersionInfo } from 'app/navigation/navigation.service';
-    Please visit <a href="https://angular.io/">angular.io</a> to see documentation for the current version of Angular.</p>
-  </div>
-  `
+    <div *ngIf="mode === 'archive'" class="mode-banner alert archive-warning">
```
Note: the changes seem not to have messed up the component's appearance:
![Screenshot at 2021-09-03 23-45-43](https://user-images.githubusercontent.com/61631103/132071967-20ad0f23-ae95-4d76-b042-5efbb2f8f187.png)

mary-poppins commented:
You can preview dc6be68 at https://pr43218-dc6be68.ngbuilds.io/.

mary-poppins commented:
You can preview 0ad9711 at https://pr43218-0ad9711.ngbuilds.io/.

gkalpak commented on aio/src/app/app.component.ts:
```diff
@@ -23,7 +37,6 @@ export const showFloatingTocWidth = 800;
   templateUrl: './app.component.html',
 })
 export class AppComponent implements OnInit {
-
```
Out of curiosity, why this change. Did `eslint` complain about it? (If so, why was it not complaining before?)

gkalpak commented on aio/src/app/app.component.ts:
```diff
@@ -78,21 +91,25 @@ export class AppComponent implements OnInit {
   showSearchResults = false;
   searchResults: Observable<SearchResults>;
-  @ViewChildren('searchBox, searchResultsView', { read: ElementRef })
+  @ViewChildren('searchBox, searchResultsView', {read: ElementRef})
```
Out of curiosity, why this change. Did `eslint` complain about it? (If so, why was it not complaining before?)

gkalpak commented on aio/src/app/app.component.ts:
```diff
@@ -118,9 +135,9 @@ export class AppComponent implements OnInit {
     /* No need to unsubscribe because this root component never dies */
 
-    this.documentService.currentDocument.subscribe(doc => this.currentDocument = doc);
+    this.documentService.currentDocument.subscribe((doc) => (this.currentDocument = doc));
```
WAT? I would very strongly prefer to not require parenthesis around single arrow function arguments 😞
Is this the default `eslint` setting? (If so, why didn't it complain before?)

gkalpak commented on aio/src/app/app.component.ts:
```diff
@@ -147,51 +164,55 @@ export class AppComponent implements OnInit {
-      this.navigationService.currentNodes,   // ...needed to determine `sidenav` state
-    ]).pipe(first())
+      this.documentService.currentDocument, // ...needed to determine host classes
+      this.navigationService.currentNodes, // ...needed to determine `sidenav` state
```
Out of curiosity, why this change. Did `eslint` complain about it? (If so, why was it not complaining before?)

dario-piotrowicz replied:
To this and the rest of **If so, why was it not complaining before?** questions...

That's really bizarre indeed... we just migrated aio to eslint recently and the only changes made to the `.eslintrc` are the addition of `@typescript-eslint/no-unused-expressions` and `max-len` :thinking: 

The only thing that I can think of is that something changed in the eslint side of things... I can see that the Renovate Bot updated a few eslint related packages in the last few weeks... I wonder if the new versions tweak some default behaviours :thinking: (seem rather unlikely but I can't think of anything else that could have caused a different behaviour...)

dario-piotrowicz replied:
actually... no.... I just realized.... I didn't update my node modules locally... so yeah... it can't be that :man_shrugging:

gkalpak replied:
This is getting confusing and unmanageably large (and there is still so work to do) :grin:
Could you please move the AIO-related bits to a separate PR (and leave only stuff related to the docs examples in this PR)?

(Alternatively, you could just revert the AIO changes from this PR and just focus on the docs examples for now. Then we can make improvements to AIO linting once this is merged. Whatever you prefer.)

dario-piotrowicz replied:
mh... yeah totally a good point! :+1: 

I'm on it!

dario-piotrowicz replied:
@gkalpak about the different formats.... I think I understood what happened.... I got a little help from prettier on the refactoring of some lines and I think it caused this changes :facepalm: 

sorry!!!! :bow: :bow: :bow: :bow: :bow: :bow:

dario-piotrowicz replied:
@gkalpak I dropped the max-len commit entirely and created this [PR](https://github.com/angular/angular/pull/43439) instead, once we can get the 43439 merged then I can rebase this one and also apply the max-len to the examples :smiley:

gkalpak commented on aio/src/app/app.component.ts:
```diff
@@ -283,20 +316,20 @@ export class AppComponent implements OnInit {
   setPageId(id: string) {
     // Special case the home page
-    this.pageId = (id === 'index') ? 'home' : id.replace('/', '-');
+    this.pageId = id === 'index' ? 'home' : id.replace('/', '-');
```
:-1: @eslint 😛

gkalpak commented on aio/src/app/layout/mode-banner/mode-banner.component.ts:
```diff
@@ -4,11 +4,15 @@ import { VersionInfo } from 'app/navigation/navigation.service';
-  </div>
-  `
+    <div *ngIf="mode === 'archive'" class="mode-banner alert archive-warning">
+      <p>
+        This is the
+        <strong>archived documentation for Angular v{{ version?.major }}.</strong> Please visit
+        <a href="https://angular.io/">angular.io</a> to see documentation for the current version of
+        Angular.
+      </p>
```
Why not:

```suggestion
      <p>
        This is the <strong>archived documentation for Angular v{{ version?.major }}.</strong> Please visit
        <a href="https://angular.io/">angular.io</a> to see documentation for the current version of Angular.
      </p>
```

That seems to be enough to satisfy the currently set limit of 120 characters.

dario-piotrowicz commented:
@gkalpak sorry, just to recap... now that I reverted the max-len changes, what is left to do is:
 - address the tasks here: https://github.com/angular/angular/pull/43218#pullrequestreview-744253351
 - decide and implement what to do for performance
 - apply the max-len fix when https://github.com/angular/angular/pull/43439 is merged
 
 does this sound right? should be all right? :thinking:

gkalpak commented:
Sounds about right 😸
<sub>(Not necessarily in that order thought 😛)</sub>

mary-poppins commented:
You can preview 57ca957 at https://pr43218-57ca957.ngbuilds.io/.

JamesHenry commented:
Sorry for applying this on the main thread, but I couldn't seem to find the original comment.

I believe you ran into this issue at some point during this PR: https://github.com/angular-eslint/angular-eslint/issues/207

I'm pleased to say I've just merged a fix for it and it should now be available on npm in v12.5.0

IgorMinar commented:
@JamesHenry thanks for the note!

@dario-piotrowicz @gkalpak should we updating to this version in a separate PR first and then rebase this PR or how would you like to proceed? (my preference is many small PRs, but sometimes there are reasons to keep several changes in a single PR).

gkalpak commented:
A separate PR sounds good to me. @dario-piotrowicz, any objections?

dario-piotrowicz commented:
@gkalpak No of course, sounds good :slightly_smiling_face: 

Also this PR has a bunch of dependencies and things blocking it, so it just make sense to do smaller somewhat-unrelated bits separately to me :slightly_smiling_face::+1: 

___

PS: I tried looking into removing tslint in the generated examples, I didn't find too much time for that but I made some small progress, I hope to get that done hopefully this weekend :slightly_smiling_face:

mary-poppins commented:
You can preview 579f419 at https://pr43218-579f419.ngbuilds.io/.

gkalpak commented:
@dario-piotrowicz, what is the state of this PR now? Are there other smaller PRs that should be merged first? What are the remaining tasks before getting this ready to merge?

dario-piotrowicz commented:
> @dario-piotrowicz, what is the state of this PR now? Are there other smaller PRs that should be merged first? What are the remaining tasks before getting this ready to merge?

@gkalpak thanks a lot for checking, I was waiting for #43746 to be merged before commenting here :slightly_smiling_face: 

___

Anyways, let me recap your tasks (from https://github.com/angular/angular/pull/43218#pullrequestreview-744253351):

- [x] Remove aio/content/examples/tslint.json (and the tslint dependency if not needed any more). Determine if we still need aio/content/examples/tsconfig.json?
- [x] Ensure we strip off files/config related to linting (such as tslint.json files) from the generated StackBlitz examples and ZIP archives.
- [ ] Have tooling remove linting comments from code snippets.

The first two should be done (as soon as #43746 is merged) *(excluding the `tsconfig.json`, as I am not sure why removing tslint should allow us to remove it)*
(the two are actually the same thing I think, unless I am missing something)

I started looking at the last point earlier today, as you can see from [the issue I opened](https://github.com/angular/angular/issues/43788), I am not really sure about how long that can take, could be a rather quick tasks or a long one depending on what we want to do (also personally I think this PR could be merged even without that done)

___

On last thing we mustn't forget is the performance gain suggested by @JamesHenry we still need to make a decision on that and implement it:
 - [ ] improve eslint performance based on James Henry's comment 

___ 

And also I have been thinking of something, which I would like to know your opinion on, this PR replaces tslint with eslint to the examples as a whole, but I also removed the linting entirely from the examples themselves, so now when we manage to merge all the PRs we will end up with the examples which have eslint disabling comments but do not have linting, and the linting is actually done just by the `example-lint` script instead, so if a used downloads an example they will see the eslint disabling comments there but eslint won't be in sight anywhere (at least from the point of view of the user who's just downloaded the example)

I think this is not the best indeed, I would imagine we will also want to look into removing the eslint disabling comments from the generated examples? (as in the examples themselves before they get zipped)
What do you think?

> ps: I am not mentioning stackblitz as I am honestly not sure how it looks there, I just know neither tslint/eslint is included in the stackblitz anyways

gkalpak commented:
Thx for the summary, @dario-piotrowicz :100:

- _Regarding removing `aio/content/examples/tsconfig.json`:_
  This is indeed unrelated to switching from `tslint` to `eslint`. It is just another file sitting next to the removed `tslint.json` that we could clean-up (it doesn't seem to be needed any more) :grin:

- _Regarding having tooling remove linting comments from code snippets:_
  Let's move the discussion to #43788. Thx for opening that btw :heart:

- _Regarding removing `eslint`-specific comments from generated StackBlitz examples and ZIP archives:_
  You are totally right that we should not include such comments in StackBlitz/ZIPs (and that is indeed part of the "have tooling remove linting comments from code snippets" task). The good news is that this will be taken care of automatically as soon as we update the [Region parser][1] to remove the comments from code-snippets, because both the [StackBlitz builder][2] and the [Example zipper][3] use the Region parser to extract the content of each file :tada:

[1]: https://github.com/angular/angular/blob/4c00e92023c526af5e09c684c78ebbf630eb150b/aio/tools/transforms/examples-package/services/region-parser.js#L9
[2]: https://github.com/angular/angular/blob/4c00e92023c526af5e09c684c78ebbf630eb150b/aio/tools/stackblitz-builder/builder.mjs#L233
[3]: https://github.com/angular/angular/blob/4c00e92023c526af5e09c684c78ebbf630eb150b/aio/tools/example-zipper/exampleZipper.mjs#L166

mary-poppins commented:
You can preview 083a354 at https://pr43218-083a354.ngbuilds.io/.

mary-poppins commented:
You can preview 3372206 at https://pr43218-3372206.ngbuilds.io/.

mary-poppins commented:
You can preview c47ddbc at https://pr43218-c47ddbc.ngbuilds.io/.

dario-piotrowicz commented:
@gkalpak here we are! all tasks should have been addressed now!!! :tada: :partying_face: :tada: :partying_face: :tada: :partying_face: :tada: :partying_face: :tada: :partying_face: :tada: :partying_face: :tada: :partying_face: :tada: :partying_face: :tada: :partying_face: :tada: 

All that's left I believe is addressing the performance comment and we should hopefully be good to go! :rocket: 

___

For reference this is James' comment on the possible performance improvements: https://github.com/angular/angular/pull/43218#issuecomment-907072083

mary-poppins commented:
You can preview d193d69 at https://pr43218-d193d69.ngbuilds.io/.

dario-piotrowicz commented:
@gkalpak ? :cry:

gkalpak commented:
Sorry for the delay, @dario-piotrowicz 😇
I'll get to this next week 🤝

dario-piotrowicz commented:
> Sorry for the delay, @dario-piotrowicz innocent I'll get to this next week handshake

No problem, actually, sorry for bothering you :bow:, I was just starting to fear that the PR got like too old/stale and you forgot about it :sweat_smile: 

I'm happy just to know you didn't forget about it, I'll be waiting for you to have a look, no rush :smiley: 
(PS: we may want to re-check things generally to make sure all is ok, but the only thing in our todo list left to be done should be the performance improvement :slightly_smiling_face:)

gkalpak commented:
For the record, I _did_ start looking into this last week. But there is still stuff to get through. I'll post my comments tomorrow :wink:
<sub>(Sorry, again, this is taking so long.)</sub>

dario-piotrowicz commented:
> For the record, I _did_ start looking into this last week. But there is still stuff to get through. I'll post my comments tomorrow wink (Sorry, again, this is taking so long.)

Sounds good, thanks a lot for having a look! :smile: 

(PS: dang, I haven't looked too closely at this in a bit, but I thought it was nearly there (excluded the performance improvement) :sweat:)

dario-piotrowicz committed:
refactor(docs-infra): use eslint in aio's example-lint script

Instead of the deprecated tslint use eslint in the aio's example-lint
script

gkalpak committed:
fixup! refactor(docs-infra): use eslint in aio's example-lint script

gkalpak committed:
fixup! refactor(docs-infra): use eslint in aio's example-lint script

Remove unused tslint config and comments from docs examples.

mary-poppins commented:
You can preview 100cbba at https://pr43218-100cbba.ngbuilds.io/.

mary-poppins commented:
You can preview 7e7b891 at https://pr43218-7e7b891.ngbuilds.io/.

dario-piotrowicz commented:
@gkalpak sorry I didn't know there was still that much to do here :sweat:, I should have checked :sob: 

(if you would have told me I would have taken care of it :cry:)

gkalpak commented:
No worries, @dario-piotrowicz 😅

To sum up, here is what I have done:
- Rebased on master.
- Squashed previous fixup commits (to "reset" review history).
- Added a couple of fixup commits to:
  - Add a `tsconfig.json` file to speed up linting for TS files.
  - Change the eslint config to make not inherit from `aio/.eslintrc.json`, because (a) it made eslint behave in unexpected ways (i.e. some rules that we thought would be inherited and applied, weren't) and (b) I don't see a strong reason to tie the two configs together given that the docs example are entirely independent from the AIO app.
  - Make the necessary fixes for linting to pass.
  - Remove obsolete tslint bits.

I have also disabled some rules in `aio/content/examples/.eslintrc.json` that caused many errors and added comments (in case we decide to eventually fix the violations and enable them - bad that shouldn't happen as part of this PR).

So, the only remaining step (afaict) is for you, @dario-piotrowicz, to take a look at my fixup commits and check if I missed anything. Then, this should be good to go 🚀

gkalpak reviewed (approved):
LGTM

Reviewed-for: global-docs-approvers

dario-piotrowicz commented:
@gkalpak thank you so very much for going out of your way to help me with the PR! :slightly_smiling_face: (again, sorry I really didn't want to put you to work on this :cry:)

I see that you've chosen the tsconfig route then for the performance improvement :slightly_smiling_face::+1: 

I'll pull down the latest changes and will have another look (but knowing you I can't imagine I'll be able to find something :heart_eyes:) :+1:

dario-piotrowicz commented on aio/content/examples/.eslintrc.json:
```diff
@@ -1,23 +1,97 @@
+  "ignorePatterns": [
+    "*.js",
+    "styleguide/**/*.avoid.ts",
+    "upgrade-*"
```
@gkalpak why this addition? :thinking:

gkalpak replied:
Oh, right, I forgot to add a comment there. I excluded the `upgrade-*` directories because they produced many errors.
Could you add a TODO comment to fix the violations and "un-ignore" the `upgrade-*` directories? :pray:

dario-piotrowicz replied:
ah... that's too bad :sweat: :cry: , sure I'll add a TODO comment :slightly_smiling_face::+1:

dario-piotrowicz replied:
comment added :slightly_smiling_face:

dario-piotrowicz commented on aio/.eslintrc.json:
```diff
@@ -38,6 +37,7 @@
         ],
         "dot-notation": "error",
         "indent": "off",
+        "max-len": ["error" , 120],
```
@gkalpak sorry, not sure why I ended up adding it in the wrong order :cry:

gkalpak replied:
No worries, can happen to anyone :smiley:

dario-piotrowicz committed:
fixup! refactor(docs-infra): use eslint in aio's example-lint script

mary-poppins commented:
You can preview 911762e at https://pr43218-911762e.ngbuilds.io/.

dario-piotrowicz committed:
fixup! refactor(docs-infra): use eslint in aio's example-lint script

mary-poppins commented:
You can preview 2c7622b at https://pr43218-2c7622b.ngbuilds.io/.

dario-piotrowicz commented:
@gkalpak I just checked and all seems good to me! :smile: 

___

I've checked (just a couple of course checking all of them would require days :sweat_smile:) and seems like the eslint comments are indeed being successfully removed from the aio examples and stackblitz:
![Screenshot at 2021-12-05 21-54-55](https://user-images.githubusercontent.com/61631103/144765427-cca163cc-61a2-4fce-8a10-942b593bf508.png)
![Screenshot at 2021-12-05 21-44-02](https://user-images.githubusercontent.com/61631103/144765152-67701efd-4ddf-4acc-9e32-06f86fcee55f.png)
![Screenshot at 2021-12-05 21-45-08](https://user-images.githubusercontent.com/61631103/144765146-e8f3bb1c-801f-4c0a-a51e-b557affaf871.png)

(PS, this is currently how it is live (I didn't even notice this before) :fearful::
![Screenshot at 2021-12-05 21-56-02](https://user-images.githubusercontent.com/61631103/144765470-f6d4ade6-9dd7-4dd3-9ed6-c85fe4121dea.png)
)
\
\
![Screenshot at 2021-12-05 21-41-18](https://user-images.githubusercontent.com/61631103/144765183-9da11cd1-3b5a-4e04-814b-0b5bcb93f264.png)
![Screenshot at 2021-12-05 21-41-04](https://user-images.githubusercontent.com/61631103/144765191-02e4ae61-c91a-4b1d-835d-d19ea9a1603d.png)

> (well, we did check this in the other PR anyways... but yeah I think it was worth double checking :slightly_smiling_face:)

___

about the linting script I must admit that on my (dinosaur) machine it is unfortunately still rather slow... but it does have quite a few files to check so I guess that it's understandable (maybe after this PR we could also look into creating a way to run linting on a specific example and not all for local development convenience? :thinking:)

___

Anyways it looks all good from my point of view :smiley: (and again @gkalpak thank you so very much for the help :bow:)

gkalpak commented on aio/.eslintrc.json:
```diff
@@ -37,7 +37,7 @@
         "dot-notation": "error",
         "indent": "off",
-        "max-len": ["error" , 120],
+        "max-len": ["error", 120],
```
:100:

gkalpak commented:
Thx for checking, @dario-piotrowicz :heart:
And thx for doing a ton of preparatory work in other PRs for this change 🥇

##
Regarding the slow-down you experience, we should investigate whether we can improve things, but should not block this PR on that imo.

OOC, how long does it take on your machine (and how long did it take with tslint)?

jessicajaniuk reviewed (approved):
LGTM :cookie:

Thanks for this, @dario-piotrowicz! Just left a few comments regarding A11y lint rules. I'm overall excited about this. I much prefer the cleanliness of the string eslint rules to the tslint ones. Great work!

reviewed-for: public-api

gkalpak commented on aio/content/examples/template-syntax/src/app/hero-detail.component.ts:
```diff
@@ -1,17 +1,16 @@
   styles: ['button {margin-left: 8px} div {margin: 8px 0} img {height:24px}'],
   template: `
   <div>
+    <!-- eslint-disable-next-line @angular-eslint/template/accessibility-alt-text -->
```
Actually, I think we can remove this comment here, since we have disabled the rule.
(But we do have a TODO to fix violations and re-enable the rule :grin:)

Same for the other comments below.

dario-piotrowicz commented:
> Thx for checking, @dario-piotrowicz heart And thx for doing a ton of preparatory work in other PRs for this change 1st_place_medal
> 
> Regarding the slow-down you experience, we should investigate whether we can improve things, but should not block this PR on that imo.
> 
> OOC, how long does it take on your machine (and how long did it take with tslint)?

Yeah about the slow-down I think it does make sense to not having that block this PR :slightly_smiling_face: 

By the way, this is how long `yarn example-lint` takes me on my machine :sweat: :
![Screenshot at 2021-12-06 20-35-25](https://user-images.githubusercontent.com/61631103/144918849-f304d7c4-28a5-476e-aaa5-b6d30ac9500b.png)

And this is on master:
![Screenshot at 2021-12-06 20-36-52](https://user-images.githubusercontent.com/61631103/144918928-6c96ef00-e20e-4a5e-915e-0b3df5fd4b55.png)

:roll_eyes:

dario-piotrowicz committed:
fixup! refactor(docs-infra): use eslint in aio's example-lint script

dario-piotrowicz commented:
Thank you very much @jessicajaniuk , as @gkalpak noted the eslint disable comments were actually no longer necessary as the are turned off in the eslintrc and there's a TODO there, so I just removed them :slightly_smiling_face: 

I also noticed we had more no-longer necessary eslint disable comments, thanks a bunch for making me notice :smiley:

dario-piotrowicz committed:
fixup! refactor(docs-infra): use eslint in aio's example-lint script

mary-poppins commented:
You can preview 75c9035 at https://pr43218-75c9035.ngbuilds.io/.

dario-piotrowicz commented:
@gkalpak I just also removed a few eslint disable comments there were left in the upgrade-* directories :slightly_smiling_face:

mary-poppins commented:
You can preview 558051c at https://pr43218-558051c.ngbuilds.io/.

gkalpak reviewed (commented):
Thx for making the changes, @dario-piotrowicz!

Regarding the slowdown:
That difference (10x) is unexpected. On my machine, tslint is still faster but the difference is much smaller: 15s (tslint) vs 38s (eslint) 🤷‍♂️

dario-piotrowicz commented on aio/src/app/custom-elements/contributor/contributor.component.ts:
```diff
@@ -8,7 +8,6 @@ import { CONTENT_URL_PREFIX } from 'app/documents/document.service';
   template: `
     <div [ngClass]="{ 'flipped': person.isFlipped }" class="contributor-card">
 
-        <!-- eslint-disable-next-line @angular-eslint/template/click-events-have-key-events -->
```
damn! that's not in the examples code! :sweat: 

sorry my bad! I'll remove it :bow: :cry:

dario-piotrowicz replied:
yeah the aio test is indeed failing, sorry I totally missed it :sweat:

dylhunn reviewed (approved):
reviewed-for: public-api

dario-piotrowicz committed:
fixup! refactor(docs-infra): use eslint in aio's example-lint script

mary-poppins commented:
You can preview 8420d60 at https://pr43218-8420d60.ngbuilds.io/.

dario-piotrowicz commented:
@gkalpak what do you think, is this one ready to go? :smiley:

gkalpak commented:
It still needs one more `public-api` approval. Then it's good to go 🚀

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

Files reviewed: aio/content/examples/styleguide/*

alxhub commented:
This PR was merged into the repository by commit 835987b78bb28463a780cdc6f4fe0f68c8c93983.

AndrewKushnir commented:
@dario-piotrowicz just wanted to say Thank You for all your work on this PR (including communication and addressing the feedback), this is very impressive (150 files changed)! 👍

dario-piotrowicz commented:
> @dario-piotrowicz just wanted to say Thank You for all your work on this PR (including communication and addressing the feedback), this is very impressive (150 files changed)! +1

Thanks a lot @AndrewKushnir! :heart: 

But I can take all the credit, @gkalpak helped me a lot, and even added a couple of commits for me too :heart_eyes: 
(also everyone who chipped in and reviewed the PR has been of great help as well :relaxed:)

The cool thing is, I think (I may need to double check), that with this PR merged all instances of tslint have been removed from all of the aio project (including the site, tooling and here the examples) :smiley: :partying_face: :rocket: :sunglasses: :tada:
